### PR TITLE
week 4 assessment complete

### PR DESCRIPTION
**Pull request by jumba23**
# Submitting Week 4 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [ X] JavaScript Coding Challenges
- [ X] Ruby Coding Challenges
- [ X] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -10,41 +10,157 @@
...
 // b) Create the function that makes the test pass.
 
+// =============================== <<<<< PSEUDO CODE >>>>> ===========================================
+// our function will take an array as a parameter
+// first step is removing element at index[0]. For this we can use .shift method
+// then we are going to use a loop do shuffle our array
+// we are going to need few helper variables to iterate over the array and swap the values (or in our case, shuffle the entire array)
+// we can use more than one for loop here but we are also going to need Math method to generate random numbers we need in order to shuffle indexes in our case
+//by shuffling indexes we are going to shuffle values as well
+// im leaning toward while loop since I haven't used it in a while.
```
Comment (elysemontano) [inline code-challenges.test.js:60]: Excellent pseudocode and I like that you are stretching yourself to try an approach you haven't used in a while.  Great work!

File: code-challenges.test.js [hunk 2]
```diff
@@ -10,41 +10,157 @@
...
+    const votes2 = { upVotes: 2, downVotes: 33 };
+    // Expected output: -31
+    const votes3 = { upVotes: 41, downVotes: 0 };
+    // Expected output: 41
+    const votes4 = { upVotes: 0 };
+    // Expected output: "No key named downVotes"
+    const votes5 = { upVotes: 0, downVotes: 21 };
+    // Expected output: -21
+    const votes6 = { downVotes: 17 };
+    // Expected output: "No key named upVotes"
```
Comment (elysemontano) [inline code-challenges.test.js:97]: Awesome edge cases!!

File: code-challenges.test.js [hunk 3]
```diff
@@ -10,41 +10,157 @@
...
+
+// =============================== <<<<< PSEUDO CODE >>>>> ===========================================
+//two arguments (arrays) will be passed into this function
+//first step of merging will be done with help of spread operator
+//second step will filter out duplicates. We can use .filter method with help of .indexOf or maybe new Set collection
+//return will only provide unique values, only once. We are going to use spread operator here before creating new Set collection
+
+const removeDuplicates = (arr1, arr2) => {
+  const newArray = [...arr1, ...arr2];
+  return [...newArray, new Set(newArray)];
```
Comment (elysemontano) [inline code-challenges.test.js:163]: Nice work!  Can you think of how to condense this down even further using the same logic?

File: interview-questions.md [hunk 4]
```diff
@@ -8,42 +8,42 @@ Then, research the question to expand on your answer. Even if you feel you have
...
-Researched answer:
+Researched answer: A block statement is the same thing as a method, BUT it does not belong to an object. Block does not have their own name, it only accepts arguments and returns values. In other programing languages Ruby blocks are called closures. To call a block within a method with a value we can use a "yield" statement.
 
 1. How do you extract a value in a Ruby hash?
 
-Your answer:
+Your answer: In order to access/extract a value in Ruby we use key-value approach. Which basically means, first we write the name of the hash, followed by square brackets with the key value. "Puts" command will give us the value of this key-value pair.
 
-Researched answer:
+Researched answer: in Ruby calling a hash followed by a key name within brackets grabs the value associated with that key. We can also use .extract to access specific data for example, user.extract!("email").keys # ["john@email"] where we want to access emails with value of "john@email". Extract can be further chained with .key or .values, depending on our task at hand.
```
Comment (elysemontano) [inline interview-questions.md:37]: Awesome answers in this section!  👏

File: ruby-challenges.rb [hunk 5]
```diff
@@ -10,7 +10,13 @@
 num3 = 221
 # Expected output: '221 is odd'
 
+def even_or_odd (num)
+    if num.odd?  then p "#{num} is odd" else p "#{num} is even" end
```
Comment (elysemontano) [inline ruby-challenges.rb:14]: Excellent work here!  Something to note, oftentimes it is best practice to print when invoking the method and not within the method itself.

File: ruby-challenges.rb [hunk 6]
```diff
@@ -22,6 +28,13 @@
 beatles_album3 = 'Abbey Road'
 # Expected output: 'bby Rd'
 
+def remov_vowels (beatles_album)
+    beatles_album.delete "aeiou"
```
Comment (elysemontano) [inline ruby-challenges.rb:32]: Logic looks great, this may not handle the last variable properly though.

File: ruby-challenges.rb [hunk 7]
```diff
@@ -32,3 +45,15 @@
 # Expected output: 'LEARN is not a palindrome'
 palindrome_tester3 = 'Rotator'
 # Expected output: 'Rotator is a palindrome'
+
+def is_palindrome (word)
+    if(word == word.reverse.capitalize)
```
Comment (elysemontano) [inline ruby-challenges.rb:50]: Capitalize works great for these test variables, however, what happens if the word is all lowercase?

**Review by elysemontano — APPROVED** [review]
Awesome work Goran!  I left some comments throughout the assessment to consider.  You have incredible interview answers and your pseudocode is also really thorough.  Keep thinking about edge cases even in Ruby.
